### ISSUE TITLE
MSGraphClient inside an SPFx webpart is now breaking with random errors

### ISSUE THREAD

**Issue by shaipetel** [SharePoint/sp-dev-docs#7654]
### Target SharePoint environment

SharePoint Online

### What SharePoint development model, framework, SDK or API is this about?

💥 SharePoint Framework

### Developer environment

Windows

### What browser(s) / client(s) have you tested

- [ ] 💥 Internet Explorer
- [X] 💥 Microsoft Edge
- [X] 💥 Google Chrome
- [ ] 💥 FireFox
- [ ] 💥 Safari
- [ ] mobile (iOS/iPadOS)
- [ ] mobile (Android)
- [ ] not applicable
- [ ] other (enter in the "Additional environment details" area below)

### Additional environment details

- browser version - Latest Edge and latest Chrome
- SPFx version  1.12.1
- Node.js version 10.16.0
- etc 

### Describe the bug / error

Making a call using the default graph client:

`
context.msGraphClientFactory.getClient().then(graphClient => {
//... make any call to the graph API here
            })
`
Used to work properly. The current tenant has allowed the requested endpoints that are being called, and this has been working for months (it is a part of nightly-QA test).
Over the weekend - this same request is randomly breaking up in 1 of 2 errors:

1. more commonly: ClientAuthError: The idToken is null or empty. Please review the trace to determine the root cause. Raw ID Token Value: null
1. On rare cases "Noonce mismatch" is coming back

Clearing browser cache, cookies, local/session storage, caches, indexedDB will not help. it will randomly work and stop working on the same exact request when refreshing the page.

### Steps to reproduce

1. build an spfx project that requests the current user's calendar view
> example I use: /me/calendar/calendarview?startdatetime=[start]&enddatetime=[end]
2. SPFx version 1.12.1
3. get graph client using this code:
`
context.msGraphClientFactory.getClient().then(graphClient => {
//... make any call to the graph API here
            })
`
4. use the graphClient to make the request 
` 
       this.graphClient.api(params.path).get();
`
Random errors coming back.

### Expected behavior

The graph API should return a valid result every time.

**Comment by shaipetel** [comment]
For reference, this comes back as the full error trace:
`ClientAuthError: The idToken is null or empty. Please review the trace to determine the root cause. Raw ID Token Value: null
    at t [as constructor] (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:108682)
    at new t (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:109218)
    at Function.t.createIdTokenNullOrEmptyError (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:111539)
    at new R (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:131668)
    at e.saveTokenFromHash (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:191487)
    at e.processCallBack (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:178944)
    at e.handleAuthenticationResponse (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:180259)
    at e.<anonymous> (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:174561)
    at sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:2:4362
    at Object.next (sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:2:4467)`

**Comment by shaipetel** [comment]
Second error, for reference with error trace:
`ClientAuthError: Nonce is not matching, Nonce received:  null, nonce expected : 07a86696-532b-477d-8848-4c138dfee99f.
    at t [as constructor] (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:108682)
    at new t (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:109218)
    at Function.t.createNonceMismatchError (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:110087)
    at e.saveTokenFromHash (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:193985)
    at e.processCallBack (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:178944)
    at e.handleAuthenticationResponse (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:180259)
    at e.<anonymous> (chunk.vendors~msalImplicit_none_9f96491354de5d17929c.js:1:174561)
    at sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:2:4362
    at Object.next (sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:2:4467)
    at o (sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:2:3238)
noFH.ConsoleLogger.logWithLevel @ consolelogger.ts:220
noFH.ConsoleLogger.error @ consolelogger.ts:152
(anonymous) @ GraphHelper.modified.ts:197
(anonymous) @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
U @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
T @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
L @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
h @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
characterData (async)
S @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
c @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
M @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
(anonymous) @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
(anonymous) @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
(anonymous) @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:70
Promise.catch (async)
authProvider @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:70
e.routeResponseToCallback @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
(anonymous) @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
(anonymous) @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
e @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
e.routeResponseToPromise @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
e.sendRequestAndRouteResponse @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
e.get @ chunk.vendors~sp-http-msgraphclient_none_a688a2d883ebfeed15ce.js:1
3pZS.GraphHelper._getData @ GraphHelper.modified.ts:147
(anonymous) @ GraphHelper.modified.ts:109
Promise.then (async)
3pZS.GraphHelper.getData @ GraphHelper.modified.ts:77
(anonymous) @ DataSource.ts:200
step @ data-view-plus.js:6492
(anonymous) @ data-view-plus.js:6473
(anonymous) @ data-view-plus.js:6467
9Nh1.__awaiter @ data-view-plus.js:6463
9Nh1.DataSource.ValidateGraph @ DataSource.ts:182
(anonymous) @ DataSource.ts:68
step @ data-view-plus.js:6492
(anonymous) @ data-view-plus.js:6473
(anonymous) @ data-view-plus.js:6467
9Nh1.__awaiter @ data-view-plus.js:6463
9Nh1.DataSource.ValidateDataSource @ DataSource.ts:35
Fjw0.DataProvider.getDataProvider @ DataProvider.ts:24
(anonymous) @ DataViewPlusWebPart.ts:542
step @ data-view-plus.js:6957
(anonymous) @ data-view-plus.js:6938
fulfilled @ data-view-plus.js:6929
Promise.then (async)
step @ data-view-plus.js:6931
(anonymous) @ data-view-plus.js:6932
A4Nq.__awaiter @ data-view-plus.js:6928
A4Nq.DataViewPlusWebPart.ProcessPropertiesChanges @ DataViewPlusWebPart.ts:481
A4Nq.DataViewPlusWebPart.render @ DataViewPlusWebPart.ts:404
(anonymous) @ DataViewPlusWebPart.ts:279
notifySubscribers @ VM5754:1415
valueHasMutated @ VM5754:1659
(anonymous) @ DataViewPlusRootViewModel.ts:350
DataViewPlusRootViewModel @ DataViewPlusRootViewModel.ts:345
(anonymous) @ VM5754:4129
createViewModel @ VM5754:4435
(anonymous) @ VM5754:4399
notifySubscribers @ VM5754:1415
(anonymous) @ VM5754:3905
(anonymous) @ VM5754:3923
(anonymous) @ VM5754:3952
tryIssueCallback @ VM5754:4062
(anonymous) @ VM5754:4083
(anonymous) @ VM5754:3952
resolveViewModel @ VM5754:4128
loadViewModel @ VM5754:4045
getFirstResultFromLoaders @ VM5754:3947
(anonymous) @ VM5754:4081
possiblyGetConfigFromAmd @ VM5754:4197
resolveConfig @ VM5754:4080
(anonymous) @ VM5754:4036
possiblyGetConfigFromAmd @ VM5754:4197
loadComponent @ VM5754:4035
getFirstResultFromLoaders @ VM5754:3947
(anonymous) @ VM5754:3922
(anonymous) @ VM5754:3952
getConfig @ VM5754:4030
getFirstResultFromLoaders @ VM5754:3947
beginLoadingComponent @ VM5754:3919
loadComponentAndNotify @ VM5754:3891
get @ VM5754:3868
ko.computed.disposeWhenNodeIsRemoved @ VM5754:4379
evaluateImmediate_CallReadThenEndDependencyDetection @ VM5754:2345
evaluateImmediate_CallReadWithDependencyDetection @ VM5754:2302
evaluateImmediate @ VM5754:2267
ko.computed.ko.dependentObservable @ VM5754:2099
init @ VM5754:4361
(anonymous) @ VM5754:3744
ignore @ VM5754:1577
(anonymous) @ VM5754:3743
arrayForEach @ VM5754:182
applyBindingsToNodeInternal @ VM5754:3729
applyBindingsToNodeAndDescendantsInternal @ VM5754:3587
ko.applyBindings @ VM5754:3821
(anonymous) @ DataViewPlusWebPart.ts:296
step @ data-view-plus.js:6957
(anonymous) @ data-view-plus.js:6938
fulfilled @ data-view-plus.js:6929
Promise.then (async)
step @ data-view-plus.js:6931
(anonymous) @ data-view-plus.js:6932
A4Nq.__awaiter @ data-view-plus.js:6928
A4Nq.DataViewPlusWebPart.onInit @ DataViewPlusWebPart.ts:221
t._onInViewport @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
t._internalFirstTimeRender @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
e._renderWebPart @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
Promise.then (async)
e.loadWebPart @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
Promise.then (async)
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
os @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
ss @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
bc @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
t.unstable_runWithPriority @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
Gi @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
hc @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
ac @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
t.unstable_runWithPriority @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
Gi @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
Qi @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
qi @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
rc @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
Wc @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
t.render @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
t._renderSingleCanvasCore @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
t._renderCanvas @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
t._tryRenderSingleCanvas @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
t._tryRenderCanvas @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
r._onFieldsLoad @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
f @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
e.processAction @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
e.dispatch @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
t._endInitialize @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
Promise.then (async)
t.componentDidMount @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
t.componentDidMount @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
ss @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
bc @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
t.unstable_runWithPriority @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
Gi @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
hc @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
ac @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
Zs @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
Bc @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
rc @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
Wc @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
t.render @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:69
t.onRender @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
t._renderCoreRouter @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
t.onRender @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
t._onRender @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
t._render @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
Promise.then (async)
e._initializeApplication @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
e._initializeApplicationWithTelemetry @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
e.startApplication @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
e.navigate @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
e.navigateToApplication @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
c @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
s @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
e._startApplication @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
e._startApplication @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
(anonymous) @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
Promise.then (async)
e.start @ sp-pages-assembly_en-us_91442da501f7343a3771f1009e5e2578.js:71
(anonymous) @ Calendar-plus---Graph-API.aspx:82
(anonymous) @ Calendar-plus---Graph-API.aspx:83
Show 140 more frames`

**Comment by shaipetel** [comment]
Possible workaround:
If I pause execution for 200ms, this code never breaks. Seems like if the request is executed too soon when the web part is loaded - this error occurs.
Noticed, if any breakpoint is set during debug it was working OK.
So for now, I modified my code to wait 200ms before running the first request, it seems the issue is gone.

This is a workaround - I would say the bug still needs to be investigated and fixed.

Hope this helps!

**Comment by shaipetel** [comment]
Found the root cause and resolved.
Leaving this here for reference.
Had a different component on the page periodically clearing session/local storage on the browser. When it ran its code - it would cause this issue randomly.

Once I disabled that code - graph client runs correctly every time.

Leaving this here for reference, but closing it as it is not an SPFx issue.

Feel free to delete if you don't see value in this post.
Thanks.
